src/mame/atari/atari400.cpp

```diff
@@ -311,362 +329,297 @@ class a400_state : public atari_common_state
-	map(0xd100, 0xd1ff).noprw();
-	map(0xd200, 0xd2ff).rw("pokey", FUNC(pokey_device::read), FUNC(pokey_device::write));
-	map(0xd300, 0xd3ff).rw(m_pia, FUNC(pia6821_device::read_alt), FUNC(pia6821_device::write_alt));
-	map(0xd400, 0xd4ff).rw(m_antic, FUNC(antic_device::read), FUNC(antic_device::write));
-	map(0xd500, 0xd7ff).noprw();
-	map(0xd800, 0xffff).rw(FUNC(a400_state::a800xl_high_r), FUNC(a400_state::a800xl_high_w));
+	map.unmap_value_high();
+	a1200xl_mem(map);
+	map(0xa000, 0xbfff).view(m_basic_view);
+	m_basic_view[0](0xa000, 0xbfff).rw(FUNC(a800xl_state::ram_r<0xa000>), FUNC(a800xl_state::ram_w<0xa000>));
+	m_basic_view[1](0xa000, 0xbfff).rom().region("maincpu", 0xa000);
 }
 
-
-void a400_state::a800xl_mem(address_map &map)
+// TODO: convert to RAM device
+void a600xl_state::a600xl_mem(address_map &map)
 {
-	map(0x0000, 0xcfff).rw(FUNC(a400_state::a800xl_low_r), FUNC(a400_state::a800xl_low_w));
-	map(0xd000, 0xd0ff).rw(m_gtia, FUNC(gtia_device::read), FUNC(gtia_device::write));
-	map(0xd100, 0xd1ff).noprw();
-	map(0xd200, 0xd2ff).rw("pokey", FUNC(pokey_device::read), FUNC(pokey_device::write));
-	map(0xd300, 0xd3ff).rw(m_pia, FUNC(pia6821_device::read_alt), FUNC(pia6821_device::write_alt));
-	map(0xd400, 0xd4ff).rw(m_antic, FUNC(antic_device::read), FUNC(antic_device::write));
-	map(0xd500, 0xd7ff).noprw();
-	map(0xd800, 0xffff).rw(FUNC(a400_state::a800xl_high_r), FUNC(a400_state::a800xl_high_w));
+	map.unmap_value_high();
+	map(0x0000, 0x3fff).ram();
+	map(0x5000, 0x57ff).view(m_selftest_view);
+	m_selftest_view[0](0x5000, 0x57ff).noprw();
+	m_selftest_view[1](0x5000, 0x57ff).rom().region("maincpu", 0xd000);
+	map(0xa000, 0xbfff).view(m_basic_view);
+	m_basic_view[0](0xa000, 0xbfff).noprw();
+	m_basic_view[1](0xa000, 0xbfff).rom().region("maincpu", 0xa000);
+	map(0xc000, 0xffff).view(m_kernel_view);
+//  m_kernel_view[0](0xc000, 0xffff).rw(FUNC(a600xl_state::ram_r<0xc000>), FUNC(a600xl_state::ram_w<0xc000>));
+	m_kernel_view[0](0xc000, 0xffff).noprw();
+	m_kernel_view[1](0xc000, 0xffff).rom().region("maincpu", 0xc000);
+	map(0xd000, 0xd7ff).m(*this, FUNC(a600xl_state::hw_iomap));
 }
 
-
-void a400_state::a130xe_mem(address_map &map)
-{
-	map(0x0000, 0xcfff).rw(FUNC(a400_state::a130xe_low_r), FUNC(a400_state::a130xe_low_w));
-	map(0xd000, 0xd0ff).rw(m_gtia, FUNC(gtia_device::read), FUNC(gtia_device::write));
-	map(0xd100, 0xd1ff).noprw();
-	map(0xd200, 0xd2ff).rw("pokey", FUNC(pokey_device::read), FUNC(pokey_device::write));
-	map(0xd300, 0xd3ff).rw(m_pia, FUNC(pia6821_device::read_alt), FUNC(pia6821_device::write_alt));
-	map(0xd400, 0xd4ff).rw(m_antic, FUNC(antic_device::read), FUNC(antic_device::write));
-	map(0xd500, 0xd7ff).noprw();
-	map(0xd800, 0xffff).rw(FUNC(a400_state::a800xl_high_r), FUNC(a400_state::a800xl_high_w));
+// TODO: have to repeat everything in order to avoid "Fatal error: A memory_view can be present in only one address map." happening for selftest_view ...
```

Comment:
I guess this is where composition approach wins over inheritance for this specific case, oh well.